File: flutter_app/lib/models/manga.dart
import 'package:hive/hive.dart';

part 'manga.g.dart';

@HiveType(typeId: 0)
enum MangaStatus {
  @HiveField(0)
  ongoing('Ongoing'),
  @HiveField(1)
  completed('Completed'),
  @HiveField(2)
  hiatus('Hiatus'),
  @HiveField(3)
  cancelled('Cancelled');

  final String label;
  const MangaStatus(this.label);
}

@HiveType(typeId: 1)
enum MangaSource {
  @HiveField(0)
  mangadex('MangaDex', 'book'),
  @HiveField(1)
  mangakakalot('Mangakakalot', 'menu_book'),
  @HiveField(2)
  webtoons('Webtoons', 'auto_stories'),
  @HiveField(3)
  asurascans('Asura Scans', 'local_fire_department'),
  @HiveField(4)
  custom('Custom Source', 'extension');

  final String label;
  final String iconName;
  const MangaSource(this.label, this.iconName);
}

@HiveType(typeId: 2)
class Manga extends HiveObject {
  @HiveField(0)
  final String id;

  @HiveField(1)
  String title;

  @HiveField(2)
  String author;

  @HiveField(3)
  String artist;

  @HiveField(4)
  MangaStatus status;

  @HiveField(5)
  String synopsis;

  @HiveField(6)
  String? coverUrl;

  @HiveField(7)
  MangaSource source;

  @HiveField(8)
  List<String> genres;

  @HiveField(9)
  DateTime lastUpdated;

  @HiveField(10)
  bool isFollowed;

  @HiveField(11)
  double? rating;

  @HiveField(12)
  List<String> chapterIds;

  @HiveField(13)
  String? customSourceId;

  Manga({
    required this.id,
    required this.title,
    this.author = 'Unknown',
    this.artist = 'Unknown',
    this.status = MangaStatus.ongoing,
    this.synopsis = '',
    this.coverUrl,
    this.source = MangaSource.mangadex,
    this.genres = const [],
    DateTime? lastUpdated,
    this.isFollowed = false,
    this.rating,
    this.chapterIds = const [],
    this.customSourceId,
  }) : lastUpdated = lastUpdated ?? DateTime.now();

  Manga copyWith({
    String? id,
    String? title,
    String? author,
    String? artist,
    MangaStatus? status,
    String? synopsis,
    String? coverUrl,
    MangaSource? source,
    List<String>? genres,
    DateTime? lastUpdated,
    bool? isFollowed,
    double? rating,
    List<String>? chapterIds,
    String? customSourceId,
  }) {
    return Manga(
      id: id ?? this.id,
      title: title ?? this.title,
      author: author ?? this.author,
      artist: artist ?? this.artist,
      status: status ?? this.status,
      synopsis: synopsis ?? this.synopsis,
      coverUrl: coverUrl ?? this.coverUrl,
      source: source ?? this.source,
      genres: genres ?? this.genres,
      lastUpdated: lastUpdated ?? this.lastUpdated,
      isFollowed: isFollowed ?? this.isFollowed,
      rating: rating ?? this.rating,
      chapterIds: chapterIds ?? this.chapterIds,
      customSourceId: customSourceId ?? this.customSourceId,
    );
  }
}

File: TheReader/Models/Manga.swift
import Foundation
import SwiftData

@Model
final class Manga {
    @Attribute(.unique) var id: String
    var title: String
    var author: String
    var artist: String
    var status: MangaStatus
    var synopsis: String
    var coverURL: String?
    var source: MangaSource
    var genres: [String]
    var lastUpdated: Date
    var isFollowed: Bool
    var rating: Double?

    @Relationship(deleteRule: .cascade, inverse: \Chapter.manga)
    var chapters: [Chapter] = []

    var unreadCount: Int {
        chapters.filter { !$0.isRead }.count
    }

    var latestChapter: Chapter? {
        chapters.sorted { $0.number > $1.number }.first
    }

    init(
        id: String = UUID().uuidString,
        title: String,
        author: String = "Unknown",
        artist: String = "Unknown",
        status: MangaStatus = .ongoing,
        synopsis: String = "",
        coverURL: String? = nil,
        source: MangaSource = .mangadex,
        genres: [String] = [],
        lastUpdated: Date = .now,
        isFollowed: Bool = false,
        rating: Double? = nil
    ) {
        self.id = id
        self.title = title
        self.author = author
        self.artist = artist
        self.status = status
        self.synopsis = synopsis
        self.coverURL = coverURL
        self.source = source
        self.genres = genres
        self.lastUpdated = lastUpdated
        self.isFollowed = isFollowed
        self.rating = rating
    }
}

enum MangaStatus: String, Codable, CaseIterable {
    case ongoing = "Ongoing"
    case completed = "Completed"
    case hiatus = "Hiatus"
    case cancelled = "Cancelled"

    var color: String {
        switch self {
        case .ongoing: return "green"
        case .completed: return "blue"
        case .hiatus: return "orange"
        case .cancelled: return "red"
        }
    }
}

enum MangaSource: String, Codable, CaseIterable {
    case mangadex = "MangaDex"
    case mangakakalot = "Mangakakalot"
    case webtoons = "Webtoons"
    case asurascans = "Asura Scans"

    var iconName: String {
        switch self {
        case .mangadex: return "book.fill"
        case .mangakakalot: return "book.closed.fill"
        case .webtoons: return "scroll.fill"
        case .asurascans: return "flame.fill"
        }
    }
}
